### PR TITLE
Better handle Globus authentications

### PR DESCRIPTION
**Pull request by forsyth2**
## Summary

Objectives:
- Isolate code changes from #370 that specifically fix the Globus authentications, and build on those changes, if necessary.

Issue resolution:
- Resolves part of #339

Select one: This pull request is...
- [x] a bug fix: increment the patch version
- [x] a small improvement: increment the minor version
- [ ] a new feature: increment the minor version
- [ ] an incompatible (non-backwards compatible) API change: increment the major version

## Big Change

- [x] To merge, I will use "Create a merge commit". That is, this change is large enough to require multiple units of work (i.e., it should be multiple commits).

### 1. Does this do what we want it to do?

Required:
- [x] Product Management: I have confirmed with the stakeholders that the objectives above are correct and complete.
- [x] Testing: I have added at least one automated test. Every objective above is represented in at least one test.
- [x] Testing: I have considered likely and/or severe edge cases and have included them in testing.

If applicable:
- [ ] Testing: this pull request adds at least one new possible command line option. I have tested using this option with and without any other option that may interact with it.

### 2. Are the implementation details accurate & efficient?

Required:
- [x] Logic: I have visually inspected the entire pull request myself.
- [x] Logic: I have left GitHub comments highlighting important pieces of code logic. I have had these code blocks reviewed by at least one other team member.

If applicable:
- [x] Dependencies: This pull request introduces a new dependency. I have discussed this requirement with at least one other team member. The dependency is noted in `zstash/conda`, not just an `import` statement.

### 3. Is this well documented?

Required:
- [x] Documentation: by looking at the docs, a new user could easily understand the functionality introduced by this pull request.
  - No docs have to be updated, but the list of steps on #339 could be shortened.

### 4. Is this code clean?

Required:
- [x] Readability: The code is as simple as possible and well-commented, such that a new team member could understand what's happening.
- [x] Pre-commit checks: All the pre-commits checks have passed.

If applicable:
- [x] Software architecture: I have discussed relevant trade-offs in design decisions with at least one other team member. It is unlikely that this pull request will increase tech debt.

### CODE REVIEW

**Review by forsyth2 — COMMENTED** [review]
@TonyB9000 I've isolated code specific to Globus auths from the refactor PR (#370) to here. (I still think the refactor is important for code clarity & future development, but it is perhaps most useful to pull out this bug fix into a separate PR)

<details>

<summary> My most recent run, using the code on this branch </summary>

```bash
cd ~/ez/zstash
lcrc_conda # My function to activate conda locally
rm -rf build
conda clean --all --y
conda env create -f conda/dev.yml -n zstash-339-20250730
conda activate zstash-339-20250730
pre-commit run --all-files
python -m pip install .

# Try 1: Using whatever state my auths happened to be in
cd ../
mkdir zstash_test339_20250730_try1
cd zstash_test339_20250730_try1
mkdir zstash_demo; echo 'file0 stuff' > zstash_demo/file0.txt
zstash create --hpss=globus://6bdc7956-fc0f-4ad2-989c-7aa5ee643a79/global/homes/f/forsyth/zstash/tests/manual_run_issue339_20250730 zstash_demo
# INFO: /home/ac.forsyth2/.globus-native-apps.cfg exists. If this file does not have the proper settings, it may cause a TransferAPIError (e.g., 'Token is not active', 'No credentials supplied')
# >>> Log into NERSC, paste auth code once
# Worked!


# Try 2: Completely clean slate: no globus cfg, no globus consents
rm ~/.globus-native-apps.cfg
cd ../
mkdir zstash_test339_20250730_try2
cd zstash_test339_20250730_try2
# globus.org > File Manager > select "LCRC Improv DTN", "NERSC Perlmutter"
# https://auth.globus.org/v2/web/consents > Manage Your Consents > Globus Endpoint Performance Monitoring > rescind all"
mkdir zstash_demo; echo 'file0 stuff' > zstash_demo/file0.txt
zstash create --hpss=globus://6bdc7956-fc0f-4ad2-989c-7aa5ee643a79/global/homes/f/forsyth/zstash/tests/manual_run_issue339_20250730 zstash_demo
# INFO: /home/ac.forsyth2/.globus-native-apps.cfg does not exist. zstash will need to prompt for authentications twice, and then you will need to re-run.
# >>> Log into Argonne, NERSC login prompt (but it remembers credentials), paste auth code
# Worked!

# Try 3: Try again with the new auth state
cd ../
mkdir zstash_test339_20250730_try3
cd zstash_test339_20250730_try3
mkdir zstash_demo; echo 'file0 stuff' > zstash_demo/file0.txt
zstash create --hpss=globus://6bdc7956-fc0f-4ad2-989c-7aa5ee643a79/global/homes/f/forsyth/zstash/tests/manual_run_issue339_20250730 zstash_demo
# INFO: /home/ac.forsyth2/.globus-native-apps.cfg exists. If this file does not have the proper settings, it may cause a TransferAPIError (e.g., 'Token is not active', 'No credentials supplied')
# >>> No prompt to log into Argonne/NERSC, paste auth code once
# Worked!

# Try 4: Try again from the same directory, as if try 3 was the "toy problem"
cd zstash_test339_20250730_try3
zstash create --hpss=globus://6bdc7956-fc0f-4ad2-989c-7aa5ee643a79/global/homes/f/forsyth/zstash/tests/manual_run_issue339_20250730 zstash_demo
# INFO: /home/ac.forsyth2/.globus-native-apps.cfg exists. If this file does not have the proper settings, it may cause a TransferAPIError (e.g., 'Token is not active', 'No credentials supplied')
# >>> No prompt to log into Argonne/NERSC, paste auth code once
# Worked! But still needed the auth code!!
```

</details>

Referring to the list of Globus problems on #339:

1. Login to Globus web interface and activate end points -- This was always needed.
2. Delete existing globus cfg file -- it really doesn't seem to matter if we delete this or not, the current code will always ask for exactly one auth code. It does look like it's almost never in the correct configuration, so maybe we can just always auto-delete it before each try.
3. Start interactive zstash test transfer -- Now skippable
4. Start a second interactive zstash transfer -- Now skippable
5. Start long transfer -- Got better: we only ever have to paste the auth code once now and we never have to do a "toy problem" first before running what we really want to run. Got worse: we appear to _always_ need an auth code (this will pose a problem for situations where we run a toy problem before running a script that calls `zstash`: the script would now hang waiting for an auth code). (Basically, now instead of entering an auth code 2 times or 0 times, it's always 1 time).
6. Using Globus web interface, manually transfer zstash files that were not transferred due to token expiration. -- I don't think this is resolved...
7. Repeat steps (2) to (4) above. Restart zstash archiving that stopped

**Comment by forsyth2** [comment]
> (Basically, now instead of entering an auth code 2 times or 0 times, it's always 1 time).

Claude suggests:
> This is actually a security feature - Globus wants explicit user consent for accessing specific endpoints, even if you have general transfer permissions.
> The "always need auth code once per session" behavior is likely intentional for this type of endpoint-specific access pattern.

File: zstash/globus_utils.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+    # Get the required scopes
+    all_scopes = get_all_endpoint_scopes(both_endpoints)
+
+    # Use the NativeAppAuthClient pattern from the documentation
+    client = NativeAppAuthClient(ZSTASH_CLIENT_ID)
+    client.oauth2_start_flow(
+        requested_scopes=all_scopes,
+        refresh_tokens=True,  # This is the key to persistent auth!
+    )
```
Comment (forsyth2) [inline zstash/globus_utils.py:154]: This addresses #338 -- the `oauth2_start_flow` comes from https://globus-sdk-python.readthedocs.io/en/stable/examples/native_app.html#with-refresh-tokens.

Relatedly, I believe this is the Globus Flow that @rljacob mentioned earlier today.

File: zstash/globus_utils.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+import json
+import os
+import os.path
+import re
+import socket
+import sys
+from typing import Dict, List, Optional
+
+from globus_sdk import (
+    NativeAppAuthClient,
```
Comment (forsyth2) [inline zstash/globus_utils.py:13]: Now we're using the Globus client instead of the `fair_research_login` one, which means 1) it will be easier to ask the Globus support team for help in the future, and 2) we can close #253 as irrelevant.

File: tests/scripts/globus_auth.bash [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+check_log_does_not_have()
+{
+    local not_expected_grep="${1}"
+    local log_file="${2}"
+    grep "${not_expected_grep}" ${log_file}
+    if [ $? == 0 ]; then
+        echo "Not-expected grep '${expected_grep}' was found in ${log_file}. Test failed."
+        exit 2
+    fi
+}
```
Comment (forsyth2) [inline tests/scripts/globus_auth.bash:39]: I tried out a similar testing setup for #378. I think these bash tests with `grep` checks are a better testing method for `zstash` than the existing "unit" tests (which in reality are closer to integration tests, as they do quite a few system calls with `tests/base.run_cmd`...)

File: zstash/globus_utils.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+# The functions here don't rely on the global variables defined in globus.py.
+
+
+# Primarily used by globus_activate ###########################################
+def check_state_files():
+    if os.path.exists(GLOBUS_CFG):
+        logger.warning(
+            f"Globus CFG {GLOBUS_CFG} exists. This may be left over from earlier versions of zstash, and may cause issues. Consider deleting."
+        )
+
```
Comment (forsyth2) [inline zstash/globus_utils.py:57]: We no longer need to worry about `GLOBUS_CFG`, resolving `(2) Delete existing globus cfg file` from #339

File: zstash/globus_utils.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        token_data = tokens["transfer.api.globus.org"]
+        if "refresh_token" in token_data:
+            logger.info("Found stored refresh token - using it")
+            # Create a simple auth client for the RefreshTokenAuthorizer
+            auth_client = NativeAppAuthClient(ZSTASH_CLIENT_ID)
+            transfer_authorizer = RefreshTokenAuthorizer(
+                refresh_token=token_data["refresh_token"], auth_client=auth_client
+            )
+            transfer_client = TransferClient(authorizer=transfer_authorizer)
+            return transfer_client
```
Comment (forsyth2) [inline zstash/globus_utils.py:141]: This resolves `(3) Start interactive zstash test transfer` & `(4) Start a second interactive zstash transfer`, and  from #339, as we no longer need to run a toy problem first. The only issue is that the very first run must be authenticated into. There's really no way of getting around that.

File: zstash/globus_utils.py (hunk 1 above)
Comment (forsyth2) [reply zstash/globus_utils.py:154]: I believe this `oauth2_start_flow` may also resolve the `(5) Start long transfer [...] Note that this transfer is limited to 48 hours due to Globus token expiration.` and `(6) Using Globus web interface, manually transfer zstash files that were not transferred due to token expiration.` of #339.

(Previously, that is, on `main`, we were using `native_client.login(no_local_server=True, refresh_tokens=True)`, so my hope is that the `oauth2_start_flow` is nicer to use).

But @TonyB9000 I think I'd need you to test on a long run to be sure that this is in fact the case.

File: zstash/globus_utils.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            logger.error("Consent required - this suggests scope issues.")
+            logger.error(
+                "With proper scope handling, this block should not be reached."
+            )
+            logger.error(
+                "Please report this bug at https://github.com/E3SM-Project/zstash/issues, with details of what you were trying to do."
+            )
+            raise RuntimeError(
+                "Insufficient Globus consents - please report this bug"
+            ) from err
```
Comment (forsyth2) [inline zstash/globus_utils.py:270]: I left a check in this block, but it's a little unclear if/how it would be reached now, so I figured it would be best to just leave a debugging note here.

**Review by forsyth2 — COMMENTED** [review]
@TonyB9000 Ok, I think I've finally got a working version! This actually allows users to:
1. Authenticate on the first new run (no toy problem needed)
2. Do future runs without any further authentications.

This can be replace a number of PRs/commits, and resolve a number of open issues:
- This PR moves the changes from the prototype script of #382 into production code (i.e., the actual `zstash` code base).
- This PR replaces the commits https://github.com/E3SM-Project/zstash/pull/370/commits/9edaf6ff94b37aab3a71b80df8aee9233f6eed3f and https://github.com/E3SM-Project/zstash/pull/370/commits/747feb292f988498fa728b478f4703326504ec47 in #370. Again, I still want to merge the refactor to clean up the code base, but I think it makes sense to pull out these auth changes into their own PR so we can test with more modularity.
- See inline comments for more issue/PR resolutions.

I still need to check how the existing unit tests run after all these changes, but my test of the problematic functionality has been successful.

File: zstash/globus_utils.py (hunk 1 above)
Comment (TonyB9000) [reply zstash/globus_utils.py:154]: @forsyth2 This is great.  Hopefully (using the new CMIP6 production workflow) we only need to fetch an archive "as-needed", as opposed to fetching them all at once - so each fetch should last less than a day - I hope.

(I still need to understand how the last fetch (wlin's NERSC directory) failed.  I'll try it again soon.

**Comment by forsyth2** [comment]
Latest commits are rebased of `main` (that is, they include the #378 merge)

**Review by forsyth2 — COMMENTED** [review]
The test `tests/scripts/globus_auth.bash` confirms we only have to do one authentication now. I've also updated that script to replace the Globus unit test, which was becoming tech debt for two reasons -- 1) it essentially duplicated the `activate` function and other internal workings, meaning the actual code for those parts wasn't getting tested, and 2) it relied on an older version of the `Client`.

This PR addresses steps 2-4 of #339. That is, users will no longer have to handle `~/.globus-native-apps.cfg` and run a "toy" problem first. Step 6 (the early token expiration) remains an open problem, however, and will need to be addressed in a future pull request.

**Comment by forsyth2** [comment]
> Will you issue a release soon? 

That will be release candidate `rc1` once we begin the Unified testing period at the beginning of October, and then eventually an actual release. In the meantime, you can use `zstash` from a `dev` environment rather than from Unified.
